native/dlarf.go

```diff
@@ -0,0 +1,70 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+// Dlarf applies an elementary reflector to a general rectangular matrix c
+// This computes
+//  c = h * c if side == Left
+//  c = c * h if side == right
+// where
+//  h = 1 - tau * v * v^T
+// and c is an m * n matrix.
+//
+// Work is pre-allocated memory of size at least m if side == Left and at least
+// n if side == Right.
+func (impl Implementation) Dlarf(side blas.Side, m, n int, v []float64, incv int, tau float64, c []float64, ldc int, work []float64) {
+	applyleft := side == blas.Left
+	if (applyleft && len(work) < n) || (!applyleft && len(work) < m) {
+		panic("dlarf: insufficient work length")
+	}
+	lastv := 0 // last non-zero element of v
+	lastc := 0 // last non-zero row/column of c
+	if tau != 0 {
+		var i int
+		if applyleft {
+			lastv = m
+		} else {
+			lastv = n
+		}
+		if incv > 0 {
+			i = (lastv - 1) * incv
+		}
+		// Look for the last non-zero row in v
+		for lastv >= 0 && v[i] == 0 {
+			lastv--
+			i -= incv
+		}
+		if applyleft {
+			// Scan for the last non-zero column in C(0:lastv, :)
```

Comment:
It is matlab notation, but it is also the notation in the FORTRAN comments. I'll change them to brackets, as that is also more go-like.